ces_admin/views.py

```diff
@@ -22,6 +33,80 @@ def ces_logout(request):
     logout(request)
     return HttpResponseRedirect('/')
 
-@login_required(login_url='/ces-login/')
-def ces_admin(request):
-    return render(request, 'ces_admin/ces-admin.html')
\ No newline at end of file
+class DashboardView(TemplateView):
+    template_name = 'ces_admin/ces-dashboard.html'
+    
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        
+        base_query = '''
+            SELECT message.text, count(state.name) as count, state.name
+            FROM decisiontree_session as session
+            JOIN decisiontree_treestate as state
+            ON session.{state_type}=state.id
+            JOIN decisiontree_message as message
+            ON state.message_id=message.id
+            {where_clause}
+            GROUP BY state.name, message.text
+            ORDER BY count DESC
+        '''
+
+        one_day_ago = datetime.datetime.now() - datetime.timedelta(days=1)
+
+        with connection.cursor() as cursor:
+            # Sessions that are in progress
+            context['open_sessions'] = Session.objects.filter(state_id__isnull=False, last_modified__gte=one_day_ago).count()
+
+            query_chart = base_query.format(state_type='state_id',
+                                            where_clause="WHERE session.state_id is not null " +
+                                                         "AND session.last_modified >= (NOW() - INTERVAL '24 hour')")
+
+            open_sessions_chart, open_sessions_map = prepare_data(cursor, query_chart)
+            context['open_sessions_chart'] = json.dumps(open_sessions_chart)
+            context['open_sessions_map'] = json.dumps(open_sessions_map)
+
+            # Sessions that the user either canceled (e.g., by typing "end") or abandoned 24 hours after starting.
+            context['canceled_sessions'] = Session.objects.filter(Q(state_id__isnull=True, canceled=True) | Q(state_id__isnull=False, last_modified__lt=one_day_ago)).count()
+
+            query_chart = base_query.format(state_type='state_at_close_id',
+                                            where_clause="WHERE (session.state_id is null " +
+                                                         "AND session.canceled=True) " +
+                                                         "OR (session.state_id is not null " +
+                                                         "AND session.last_modified < (NOW() - INTERVAL '24 hour'))")
+
+            canceled_sessions_chart, canceled_sessions_map = prepare_data(cursor, query_chart)
+            context['canceled_sessions_chart'] = json.dumps(canceled_sessions_chart)
```

Comment:
seems like you cold move the json dumping into prepare_data because that's what always happens, right?

Reply:
oh, of course